File: navigation/app/src/main/java/com/mapbox/services/android/navigation/testapp/NavigationApplication.java
package com.mapbox.services.android.navigation.testapp;

import android.app.Application;
import android.text.TextUtils;
import android.util.Log;

import com.mapbox.mapboxsdk.Mapbox;
import com.squareup.leakcanary.LeakCanary;

import timber.log.Timber;

/**
 * Created by antonio on 4/14/17.
 */

public class NavigationApplication extends Application {

  private static final String LOG_TAG = NavigationApplication.class.getSimpleName();

  @Override
  public void onCreate() {
    super.onCreate();

    // Leak canary
    if (LeakCanary.isInAnalyzerProcess(this)) {
      return;
    }
    LeakCanary.install(this);

    if (BuildConfig.DEBUG) {
      Timber.plant(new Timber.DebugTree());
    }

    // Set access token
    String mapboxAccessToken = Utils.getMapboxAccessToken(getApplicationContext());
    if (TextUtils.isEmpty(mapboxAccessToken)) {
      Log.w(LOG_TAG, "Warning: access token isn't set.");
    }

    Mapbox.getInstance(getApplicationContext(), mapboxAccessToken);
  }

}

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/NavigationApplication.kt
package com.mapbox.services.android.navigation.testapp

import android.os.StrictMode
import android.text.TextUtils
import androidx.multidex.MultiDexApplication
import com.mapbox.android.search.MapboxSearch
import com.mapbox.android.search.MapboxSearchOptions
import com.mapbox.crashmonitor.CrashMonitor
import com.mapbox.mapboxsdk.Mapbox
import com.mapbox.services.android.navigation.testapp.example.utils.DelegatesExt
import com.squareup.leakcanary.LeakCanary
import timber.log.Timber

private const val DEFAULT_MAPBOX_ACCESS_TOKEN = "YOUR_MAPBOX_ACCESS_TOKEN_GOES_HERE"

class NavigationApplication : MultiDexApplication() {

  companion object {
    var instance: NavigationApplication by DelegatesExt.notNullSingleValue()
  }

  override fun onCreate() {
    super.onCreate()
    instance = this
    setupTimber()
    setupStrictMode()
    setupCanary()
    setupMapbox()
    setupCrashMonitor()
  }

  private fun setupTimber() {
    if (BuildConfig.DEBUG) {
      Timber.plant(Timber.DebugTree())
    }
  }

  private fun setupStrictMode() {
    if (BuildConfig.DEBUG) {
      StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder()
          .detectAll()
          .penaltyLog()
          .build())
      StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder()
          .detectAll()
          .penaltyLog()
          .build())
    }
  }

  private fun setupCanary() {
    if (LeakCanary.isInAnalyzerProcess(this)) {
      // This process is dedicated to LeakCanary for heap analysis.
      // You should not init your app in this process.
      return
    }
    LeakCanary.install(this)
  }

  private fun setupMapbox() {
    val mapboxAccessToken = Utils.getMapboxAccessToken(applicationContext)
    if (TextUtils.isEmpty(mapboxAccessToken) || mapboxAccessToken == DEFAULT_MAPBOX_ACCESS_TOKEN) {
      Timber.w("Mapbox access token isn't set!")
    }

    val cachingMode = MapboxSearchOptions().setCachingEnabled(true)
    MapboxSearch.getInstance(applicationContext, mapboxAccessToken, cachingMode)
    Mapbox.getInstance(applicationContext, mapboxAccessToken)
  }

  private fun setupCrashMonitor() {
    val crashMonitor = CrashMonitor { crashDetails ->
      throw Exception(crashDetails)
    }
    try {
      crashMonitor.monitor(applicationInfo.dataDir)
    } catch (e: Exception) {
      Timber.e("Couldn't monitor for crashes: ${e.message}")
    }
  }
}
